src/GovUk.Education.ExploreEducationStatistics.Data.Services/ViewModels/LocationViewModelBuilder.cs

```diff
@@ -0,0 +1,123 @@
+        LocationAttributeNode locationAttributeNode,
+        IReadOnlyDictionary<string, GeoJson>? geoJsonByCode)
+    {
+        var locationAttribute = locationAttributeNode.Attribute;
+        return locationAttributeNode.IsLeaf
+            ? BuildLocationAttributeLeafViewModel(locationAttributeNode, geoJsonByCode)
+            : new LocationAttributeViewModel
+            {
+                Label = locationAttribute.Name ?? string.Empty,
+                Level = locationAttribute.GeographicLevel,
+                Value = locationAttribute.GetCodeOrFallback(),
+                Options = BuildLocationAttributeViewModels(locationAttributeNode.Children, geoJsonByCode)
+            };
+    }
+
+    private static LocationAttributeViewModel BuildLocationAttributeLeafViewModel(
+        LocationAttributeNode locationAttributeNode,
+        IReadOnlyDictionary<string, GeoJson>? geoJsonByCode)
+    {
+        var locationAttribute = locationAttributeNode.Attribute;
+        var code = locationAttribute.GetCodeOrFallback();
+
+        var geoJson = code.IsNullOrEmpty()
+            ? null
+            : geoJsonByCode?.GetValueOrDefault(code)?.Deserialized;
+
+        return new LocationAttributeViewModel
+        {
+            Id = locationAttributeNode.LocationId,
+            GeoJson = geoJson,
+            Label = locationAttribute.Name ?? string.Empty,
+            Value = code
+        };
+    }
+
+    private static List<LocationAttributeViewModel> DeduplicateLocationViewModels(
+        List<LocationAttributeViewModel> locations)
+    {
+        // Find duplicates by label which also have the same value.
+        // They are unique by some other attribute of location and appending the value of this attribute won't deduplicate them.
+        // TODO EES-2954 enhance the deduplication to append other attributes of the location
+        var notDistinctByLabelAndValue = locations
+            .GroupBy(model => (model.Value, model.Label))
+            .Where(grouping => grouping.Count() > 1)
+            .SelectMany(grouping => grouping)
+            .ToList();
+
+        // Excluding those, find duplicates due to having the same label
+        var notDistinctByLabel = locations
+            .Except(notDistinctByLabelAndValue)
+            .GroupBy(model => model.Label.ToLower())
+            .Where(grouping => grouping.Count() > 1)
+            .SelectMany(grouping => grouping)
+            .ToList();
+
+        if (!notDistinctByLabel.Any())
+        {
+            return locations;
+        }
+
+        return locations.Select(location =>
+        {
+            if (notDistinctByLabel.Contains(location))
+            {
+                // Append the value to make them differentiable by label
+                location.Label += $" ({location.Value})";
+            }
+
+            return location;
+        }).ToList();
+    }
+
+    private static string OrderLocationAttributes(LocationAttributeNode node)
```

Comment:
I'd expect this method to do ordering, so I feel like the name might not be quite right or I'm not understanding it?

Reply:
Thanks, it's picking the field that is used for the ordering, so I've renamed it `OrderingKey` rather `OrderLocationAttributes`. Let me know if you think of a better name.